File: creator-games777/assets/scripts/fgame/Timer.ts
/**
 * Timer helpers for FGame on Creator (replaces Special/CocosFish2/Timer.lua).
 */
type TimerHandle = { id: number; cancelled: boolean };

const timers = new Set<TimerHandle>();
let nextId = 1;

export function StartTimer(delaySec: number, cb: () => void, loop = false): TimerHandle {
  const handle: TimerHandle = { id: nextId++, cancelled: false };
  timers.add(handle);

  const tick = () => {
    if (handle.cancelled) return;
    cb();
    if (loop && !handle.cancelled) {
      setTimeout(tick, delaySec * 1000);
    } else {
      timers.delete(handle);
    }
  };
  setTimeout(tick, delaySec * 1000);
  return handle;
}

export function StopTimer(handle: TimerHandle | null): void {
  if (!handle) return;
  handle.cancelled = true;
  timers.delete(handle);
}

export function StopAllTimer(): void {
  for (const t of timers) t.cancelled = true;
  timers.clear();
}

File: FGameCommon/src/FGame/Common/Special/Creator/Timer.lua
-- Creator timer adapter (Special/Init.lua RUNTIME_IN_CREATOR).

local allTimer = {}
local nextId = 1

local function schedule(call, interval)
    local id = nextId
    nextId = nextId + 1
    allTimer[id] = true

    local function tick()
        if not allTimer[id] then return end
        call()
        if allTimer[id] then
            -- Creator / host should replace with engine scheduler; setTimeout for hybrid shell.
            if type(setTimeout) == "function" then
                setTimeout(tick, math.floor((interval or 0) * 1000))
            end
        end
    end

    if type(setTimeout) == "function" then
        setTimeout(tick, math.floor((interval or 0) * 1000))
    else
        -- Immediate fallback for environments without timers
        call()
        allTimer[id] = nil
    end
    return id
end

function StartTimer(call, interval)
    if interval == nil or interval < 0 then
        interval = 0
    end
    return schedule(call, interval)
end

function StopTimer(handle)
    if handle == nil then return end
    allTimer[handle] = nil
end

function StartOnceTimer(call, delay)
    local handler = nil
    handler = StartTimer(function()
        StopTimer(handler)
        call()
    end, delay)
    return handler
end

function StopAllTimer()
    allTimer = {}
end
